text/0000-public-private-dependencies.md

```diff
@@ -0,0 +1,192 @@
+duplicated.  This distinction will also make it possible to relax some restrictions that
+currently exist in Cargo which sometimes prevent crates from compiling.
+
+**Q: Can public become private later?**<br>
+A: Public dependencies are public within a reachable subgraph but can become private if a
+crate stops exposing a public dependency.  For instance it is very possible to have a
+family of crates that all depend on a utility crate that provides common types which is
+a public dependency for all of them.  However your own crate only becomes a user of this
+utility crate through another dependency that itself does not expose any of the types
+from that utility crate and as such the dependency is marked private.
+
+**Q: Where is public / private defined?**<br>
+Dependencies are private by default and are made public through a `public` flag in the
+dependency in the `Cargo.toml` file.  This also means that crates created before the
+implementation of this RFC will have all their dependencies private.
+
+**Q: How is backwards compatibility handled?**<br>
+A: It will continue to be permissible to "leak" dependencies and there are even some
+use cases of this, however the compiler or cargo will emit warnings if private
+dependencies become part of the public API.  Later it might even become invalid to
+publish new crates without explicitly silencing these warnings or marking the
+dependencies as public.
+
+**Q: Can I export a type from a private dependency as my own?**<br>
+For now it will not be strictly permissible to privately depend on a crate and export
+a type from their as your own.  The reason for this is that at the moment it is not
+possible to force this type to be distinct.  This means that users of the crate might
+accidentally start depending on that type to be compatible if the user starts to depend
+on the crate that actually implements that type.
+
+# Detailed design
+[design]: #detailed-design
+
+There are a few areas that require to be changed for this RFC:
+
+* The compiler needs to be extended to understand when crate dependencies are
+  considered a public dependency
+* The `Cargo.toml` manifest needs to be extended to support declaring public
+  dependencies
+* The cargo publish process needs to be changed to warn (or prevent) the publishing
+  of crates that have undeclared public dependencies
+* crates.io should show public dependencies more prominently than private ones.
+
+## Compiler Changes
+
+The main change to the compiler will be to accept a new parameter that cargo
+supplies which is a list of public dependencies.  The compiler then emits
+warnings if it encounters private dependencies leaking to the public API of a
+crate.  `cargo publish` might change this warning into an error in its lint
+step.
+
+Additionally later on the warning can turn into a hard error in general.
+
+In some situations it can be necessary to allow private dependencies to become
```

Comment:
I rather not go into too much detail about this as this is a big unknown. I think it's largely irrelevant for this RFC anyways because we initially start out with just compiler warnings. Where to go from there will be seen by how much damage this does.